### PR TITLE
[FLINK-8841] [state] Remove HashMapSerializer and use MapSerializer instead.

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

So far we had the `MapSerializer` and the `HashMapSerializer`. The two had almost identical code and the second was only used on the `HeapStateBackend`/`FSStateBackend` when creating a `MapState`. This PR removes the `HashMapSerializer` and replaces its uses with the `MapSerializer`. It also guarantees backwards compatibility.

## Brief change log

It introduces the `MigrationUtil` as an inner class of the `InstantiationUtil`. This class contains mapping between deprecated/deleted serializers and their replacements.

Also the removal of the `HashMapSerializer` uniformizes a bit the `HeapMapState` and the `RocksDBMapState`.

## Verifying this change

Added the `HeapKeyedStateBackendSnapshotMigrationTest#testMapStateMigrationAfterHashMapSerRemoval()`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

R @StefanRRichter or @aljoscha

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 1]
```diff
@@ -221,6 +226,52 @@ protected ObjectStreamClass readClassDescriptor() throws IOException, ClassNotFo
...
+		// To add a new mapping just pick a name and add an entry as the following:
+
+		GENERIC_DATA_ARRAY_SERIALIZER(
+				"org.apache.avro.generic.GenericData$Array",
+				ObjectStreamClass.lookup(KryoRegistrationSerializerConfigSnapshot.DummyRegisteredClass.class)),
+		HASH_MAP_SERIALIZER(
+				"org.apache.flink.runtime.state.HashMapSerializer",
+				ObjectStreamClass.lookup(MapSerializer.class)); // added in 1.5
+
+		/** An internal unmoddifiable map containing the mappings between deprecated and new serialzers. */
```
Comment (yuqi1129) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: `unmoddifiable`, `serialzers ` 
spelling mistake

**Comment by bowenli86** [comment]
+1

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 2]
```diff
@@ -221,6 +226,52 @@ protected ObjectStreamClass readClassDescriptor() throws IOException, ClassNotFo
...
+		private static final Map<String, ObjectStreamClass> equivalenceMap = Collections.unmodifiableMap(initMap());
```
Comment (StefanRRichter) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: If it is static final, I suggest to use capital letter and underscores for the variable name.

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 3]
```diff
@@ -221,6 +226,52 @@ protected ObjectStreamClass readClassDescriptor() throws IOException, ClassNotFo
...
+
+		/** The full name of the class of the old serializer. */
+		private final String oldSerializer;
```
Comment (StefanRRichter) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: mabye add `name` to the variable name to show it is a string, not a serializer.

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 4]
```diff
@@ -221,6 +226,52 @@ protected ObjectStreamClass readClassDescriptor() throws IOException, ClassNotFo
...
+
+		/** The serialization descriptor of the class of the new serializer. */
+		private final ObjectStreamClass newSerializer;
```
Comment (StefanRRichter) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: Here I suggest to add `StreamClass` to the name.

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 5]
```diff
@@ -221,6 +226,52 @@ protected ObjectStreamClass readClassDescriptor() throws IOException, ClassNotFo
...
+
+		MigrationUtil(final String oldSerializer, final ObjectStreamClass newSerializer) {
```
Comment (StefanRRichter) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: the `final` in on these parameters is just noise

**Comment by kl0u** [comment]
Thanks @yuqi1129, @bowenli86 and @StefanRRichter for the reviews. I integrated your comments. If you are done with reviewing, I will merge it as soon as Travis gives the green light.

**Comment by StefanRRichter** [comment]
LGTM 👍

**Pull request closed without merging**